### PR TITLE
Onigiri - Add 3DCNN Model for Mood, Emotion, and Facial Expression Analysis

### PR DESCRIPTION
**Pull request by kwon-encored**
## Overview
This PR introduces a new **multidimensional 3D CNN model** within the Onigiri project.  
The model leverages a large-scale dataset (~18TB) capturing regressional relationships between **mood, emotion, and facial expressions**, along with **gender attributes**.  

The goal is to extend the multimodal project by enabling **mood determination** from image and face data, integrated with contextual metadata.

---

## Key Features
- **New Data Integration**
  - Added ~18TB of mass data on mood, emotion, and facial expression alongside gender.
  - Preprocessing pipeline supports sequence-based image and embedding fusion.

- **3D CNN Model Implementation**
  - Input: `data_input` (sequence of facial image tensors).
  - Auxiliary Input: `site_id_input` for contextual weather embedding.
  - Weather embedding reshaped into a **weather map** and concatenated as an additional channel.
  - Temporal-spatial Conv3D layers with ELU activations.
  - Dense fully connected layers leading to mood prediction outputs.

- **Output**
  - Predicts **mood state** given image and contextual inputs.
  - Designed to integrate seamlessly with existing multimodal architecture.

---

## Motivation
This implementation expands Onigiri’s capability:
- Moves beyond **basic sentiment analysis** to deeper **mood-level understanding**.
- Bridges the gap between **visual emotion recognition** and **context-aware multimodal inference**.
- Scales to massive datasets, aligning with the multimodal project’s growth roadmap.

---

## Next Steps
- Train and benchmark the new model on curated dataset splits.
- Compare performance against existing CNN and multimodal baselines.
- Integrate evaluation metrics for mood detection accuracy and generalization.

---

### CODE REVIEW

File: src/train_gender_classifier.py [hunk 1]
```diff
@@ -36,6 +37,69 @@
                                  grayscale=grayscale,
                                  do_random_crop=do_random_crop)
 
+# onigiri - as of 2025
+df_weather_mood = DataManager("onigiri")
```
[the remark this answers is not in the document]
Comment (kwon-encored) [reply src/train_gender_classifier.py]: ⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐

File: src/train_gender_classifier.py [hunk 2]
```diff
@@ -36,6 +37,69 @@
...
+    optimizer="adam",
+    loss="categorical_crossentropy",
+    metrics=["mae"]
+)
+
+mood_model.summary()
+
+# ---- 3) Callbacks (match the style from your example) ----
+# fill these in (same variable names you used before)
+patience = 10
```
[the remark this answers is not in the document]
Comment (kwon-encored) [reply src/train_gender_classifier.py]: ⭐

File: src/train_gender_classifier.py [hunk 3]
```diff
@@ -36,6 +37,69 @@
...
+    verbose=1,
+    save_best_only=True,
+    save_weights_only=False
+)
+
+callbacks = [model_checkpoint, csv_logger, early_stop, reduce_lr]
+
+# ---- 4A) Fit with arrays / tf.data (recommended) ----
+history = mood_model.fit(
+    all_possible_combinations_input, y_true,
```
[the remark this answers is not in the document]
Comment (kwon-encored) [reply src/train_gender_classifier.py:91]: ⭐⭐⭐⭐⭐ BUT SIMPLE COMPILE ERROR ⭐⭐⭐⭐⭐⭐⭐⭐⭐

File: src/image_emotion_gender_demo.py [hunk 4]
```diff
@@ -27,15 +28,18 @@
...
 # loading models
 face_detection = load_detection_model(detection_model_path)
 emotion_classifier = load_model(emotion_model_path, compile=False)
 gender_classifier = load_model(gender_model_path, compile=False)
+mood_classifier   = load_model(onigiri_model_path, compile=False)
 
 # getting input model shapes for inference
 emotion_target_size = emotion_classifier.input_shape[1:3]
 gender_target_size = gender_classifier.input_shape[1:3]
+mood_target_size = mood_classifier.input_shape[1:3]
```
[the remark this answers is not in the document]
Comment (kwon-encored) [reply src/image_emotion_gender_demo.py:43]: ⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐
BUT also can be achieved from the compile error

File: src/train_gender_classifier.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kwon-encored) [reply src/train_gender_classifier.py:91]: done by changing the function itself

File: src/train_gender_classifier.py [hunk 5]
```diff
@@ -36,6 +37,67 @@
                                  grayscale=grayscale,
                                  do_random_crop=do_random_crop)
 
+# onigiri - as of 2025
+df_weather_mood = pd.read_csv('../datasets/onigiri.csv')
+all_possible_combinations_input, y_true = return_emotions_mood_weather_mixer_combinations(df_weather_mood, batch_size,num_epochs,patience)
```
[the remark this answers is not in the document]
Comment (kwon-encored) [reply src/train_gender_classifier.py:42]: Wrong
